File: tags/symmetric-build-3.5.19+b.1/symmetric-build-3.1.4/symmetric-core/src/main/java/org/jumpmind/symmetric/service/impl/IncomingBatchService.java
/*
 * Licensed to JumpMind Inc under one or more contributor 
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding 
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU Lesser General Public License (the
 * "License"); you may not use this file except in compliance
 * with the License. 
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see           
 * <http://www.gnu.org/licenses/>.
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License. 
 */
package org.jumpmind.symmetric.service.impl;

import java.sql.Types;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.apache.commons.lang.StringUtils;
import org.jumpmind.db.sql.ISqlRowMapper;
import org.jumpmind.db.sql.Row;
import org.jumpmind.db.sql.UniqueKeyException;
import org.jumpmind.db.sql.mapper.DateMapper;
import org.jumpmind.symmetric.common.ParameterConstants;
import org.jumpmind.symmetric.db.ISymmetricDialect;
import org.jumpmind.symmetric.model.IncomingBatch;
import org.jumpmind.symmetric.model.IncomingBatch.Status;
import org.jumpmind.symmetric.service.IIncomingBatchService;
import org.jumpmind.symmetric.service.IParameterService;
import org.jumpmind.symmetric.util.AppUtils;
import org.jumpmind.util.FormatUtils;

/**
 * @see IIncomingBatchService
 */
public class IncomingBatchService extends AbstractService implements IIncomingBatchService {

    public IncomingBatchService(IParameterService parameterService,
            ISymmetricDialect symmetricDialect) {
        super(parameterService, symmetricDialect);
        setSqlMap(new IncomingBatchServiceSqlMap(symmetricDialect.getPlatform(),
                createSqlReplacementTokens()));
    }

    public IncomingBatch findIncomingBatch(long batchId, String nodeId) {
        return sqlTemplate.queryForObject(
                getSql("selectIncomingBatchPrefixSql", "findIncomingBatchSql"),
                new IncomingBatchMapper(), batchId, nodeId);
    }

    public int countIncomingBatchesInError() {
        return sqlTemplate.queryForInt(getSql("countIncomingBatchesErrorsSql"));
    }

    public List<IncomingBatch> findIncomingBatchErrors(int maxRows) {
        return sqlTemplate.query(
                getSql("selectIncomingBatchPrefixSql", "findIncomingBatchErrorsSql"), maxRows,
                new IncomingBatchMapper());
    }

    public void markIncomingBatchesOk(String nodeId) {
        List<IncomingBatch> batches = listIncomingBatchesInErrorFor(nodeId);
        for (IncomingBatch incomingBatch : batches) {
            incomingBatch.setErrorFlag(false);
            incomingBatch.setStatus(Status.OK);
            updateIncomingBatch(incomingBatch);
        }
    }

    public List<IncomingBatch> listIncomingBatchesInErrorFor(String nodeId) {
        return sqlTemplate.query(
                getSql("selectIncomingBatchPrefixSql", "listIncomingBatchesInErrorForNodeSql"),
                new IncomingBatchMapper(), nodeId);
    }

    public List<Date> listIncomingBatchTimes(List<String> nodeIds, List<String> channels,
            List<IncomingBatch.Status> statuses, boolean ascending) {
        if (nodeIds != null && nodeIds.size() > 0 && channels != null && channels.size() > 0
                && statuses != null && statuses.size() > 0) {
            Map<String, Object> params = new HashMap<String, Object>();
            params.put("NODES", nodeIds);
            params.put("CHANNELS", channels);
            params.put("STATUSES", toStringList(statuses));
            String sql = getSql("selectCreateTimePrefixSql",
                    containsOnlyErrorStatus(statuses) ? "listIncomingBatchesInErrorSql"
                            : "listIncomingBatchesSql", ascending ? " order by create_time"
                            : " order by create_time desc");
            return sqlTemplate.query(sql, new DateMapper(), params);
        } else {
            return new ArrayList<Date>(0);
        }
    }

    public List<IncomingBatch> listIncomingBatches(List<String> nodeIds, List<String> channels,
            List<IncomingBatch.Status> statuses, Date startAtCreateTime,
            final int maxRowsToRetrieve, boolean ascending) {
        if (nodeIds != null && nodeIds.size() > 0 && channels != null && channels.size() > 0
                && statuses != null && statuses.size() > 0) {
            Map<String, Object> params = new HashMap<String, Object>();
            params.put("NODES", nodeIds);
            params.put("CHANNELS", channels);
            params.put("STATUSES", toStringList(statuses));
            params.put("CREATE_TIME", startAtCreateTime);

            String createTimeLimiter = "";
            if (startAtCreateTime != null) {
                createTimeLimiter = " and create_time " + (ascending ? ">=" : "<=")
                        + " :CREATE_TIME";
            }

            String sql = getSql("selectIncomingBatchPrefixSql",
                    containsOnlyErrorStatus(statuses) ? "listIncomingBatchesInErrorSql"
                            : "listIncomingBatchesSql", createTimeLimiter,
                    ascending ? " order by create_time" : " order by create_time desc");

            return sqlTemplate.query(sql, maxRowsToRetrieve, new IncomingBatchMapper(), params);
        } else {
            return new ArrayList<IncomingBatch>(0);
        }
    }

    protected boolean containsOnlyErrorStatus(List<IncomingBatch.Status> statuses) {
        return statuses.size() == 1 && statuses.get(0) == IncomingBatch.Status.ER;
    }

    protected List<String> toStringList(List<IncomingBatch.Status> statuses) {
        List<String> statusStrings = new ArrayList<String>(statuses.size());
        for (Status status : statuses) {
            statusStrings.add(status.name());
        }
        return statusStrings;

    }

    public boolean acquireIncomingBatch(IncomingBatch batch) {
        boolean okayToProcess = true;
        if (batch.isPersistable()) {
            IncomingBatch existingBatch = null;

            try {
                insertIncomingBatch(batch);
            } catch (UniqueKeyException e) {
                batch.setRetry(true);
                existingBatch = findIncomingBatch(batch.getBatchId(), batch.getNodeId());
            }

            if (batch.isRetry()) {
                if (existingBatch.getStatus() == Status.ER
                        || existingBatch.getStatus() == Status.LD
                        || !parameterService
                                .is(ParameterConstants.INCOMING_BATCH_SKIP_DUPLICATE_BATCHES_ENABLED)) {
                    okayToProcess = true;
                    existingBatch.setStatus(Status.LD);
                    log.warn("Retrying batch {}", batch.getNodeBatchId());
                } else if (existingBatch.getStatus() == Status.IG) {
                    okayToProcess = false;
                    batch.setStatus(Status.OK);
                    batch.incrementIgnoreCount();
                    log.warn("Ignoring batch {}", batch.getNodeBatchId());
                } else {
                    okayToProcess = false;
                    batch.setStatus(existingBatch.getStatus());
                    batch.setByteCount(existingBatch.getByteCount());
                    batch.setDatabaseMillis(existingBatch.getDatabaseMillis());
                    batch.setNetworkMillis(existingBatch.getNetworkMillis());
                    batch.setFilterMillis(existingBatch.getFilterMillis());
                    batch.setSkipCount(existingBatch.getSkipCount() + 1);
                    batch.setStatementCount(existingBatch.getStatementCount());

                    existingBatch.setSkipCount(existingBatch.getSkipCount() + 1);
                    log.warn("Skipping batch {}", batch.getNodeBatchId());
                }
                updateIncomingBatch(existingBatch);
            }
        }
        return okayToProcess;
    }

    public void insertIncomingBatch(IncomingBatch batch) {
        if (batch.isPersistable()) {
            batch.setLastUpdatedHostName(AppUtils.getServerId());
            batch.setLastUpdatedTime(new Date());
            sqlTemplate.update(
                    getSql("insertIncomingBatchSql"),
                    new Object[] { batch.getBatchId(), batch.getNodeId(), batch.getChannelId(),
                            batch.getStatus().name(), batch.getNetworkMillis(),
                            batch.getFilterMillis(), batch.getDatabaseMillis(),
                            batch.getFailedRowNumber(), batch.getFailedLineNumber(),
                            batch.getByteCount(), batch.getStatementCount(),
                            batch.getFallbackInsertCount(), batch.getFallbackUpdateCount(),
                            batch.getIgnoreCount(), batch.getMissingDeleteCount(),
                            batch.getSkipCount(), batch.getSqlState(), batch.getSqlCode(),
                            FormatUtils.abbreviateForLogging(batch.getSqlMessage()),
                            batch.getLastUpdatedHostName(), batch.getLastUpdatedTime() },
                    new int[] { Types.NUMERIC, Types.VARCHAR, Types.VARCHAR, Types.CHAR,
                            Types.NUMERIC, Types.NUMERIC, Types.NUMERIC, Types.NUMERIC,
                            Types.NUMERIC, Types.NUMERIC, Types.NUMERIC, Types.NUMERIC,
                            Types.NUMERIC, Types.NUMERIC, Types.NUMERIC, Types.NUMERIC,
                            Types.VARCHAR, Types.NUMERIC, Types.VARCHAR, Types.VARCHAR,
                            Types.TIMESTAMP });
        }
    }

    public int deleteIncomingBatch(IncomingBatch batch) {
        return sqlTemplate.update(getSql("deleteIncomingBatchSql"),
                new Object[] { batch.getBatchId(), batch.getNodeId() }, new int[] { Types.NUMERIC,
                        Types.VARCHAR });
    }

    public int updateIncomingBatch(IncomingBatch batch) {
        int count = 0;
        if (batch.isPersistable()) {
            if (batch.getStatus() == IncomingBatch.Status.ER) {
                batch.setErrorFlag(true);
            } else if (batch.getStatus() == IncomingBatch.Status.OK) {
                batch.setErrorFlag(false);
            }
            batch.setLastUpdatedHostName(AppUtils.getServerId());
            batch.setLastUpdatedTime(new Date());
            count = sqlTemplate.update(
                    getSql("updateIncomingBatchSql"),
                    new Object[] { batch.getStatus().name(), batch.isErrorFlag() ? 1 : 0,
                            batch.getNetworkMillis(), batch.getFilterMillis(),
                            batch.getDatabaseMillis(), batch.getFailedRowNumber(),
                            batch.getFailedLineNumber(), batch.getByteCount(),
                            batch.getStatementCount(), batch.getFallbackInsertCount(),
                            batch.getFallbackUpdateCount(), batch.getIgnoreCount(),
                            batch.getMissingDeleteCount(), batch.getSkipCount(),
                            batch.getSqlState(), batch.getSqlCode(),
                            FormatUtils.abbreviateForLogging(batch.getSqlMessage()),
                            batch.getLastUpdatedHostName(), batch.getLastUpdatedTime(),
                            batch.getBatchId(), batch.getNodeId() }, new int[] { Types.CHAR,
                            Types.SMALLINT, Types.NUMERIC, Types.NUMERIC, Types.NUMERIC,
                            Types.NUMERIC, Types.NUMERIC, Types.NUMERIC, Types.NUMERIC,
                            Types.NUMERIC, Types.NUMERIC, Types.NUMERIC, Types.NUMERIC,
                            Types.NUMERIC, Types.VARCHAR, Types.NUMERIC, Types.VARCHAR,
                            Types.VARCHAR, Types.TIMESTAMP, Types.NUMERIC, Types.VARCHAR });
        }
        return count;
    }

    class IncomingBatchMapper implements ISqlRowMapper<IncomingBatch> {
        public IncomingBatch mapRow(Row rs) {
            IncomingBatch batch = new IncomingBatch();
            batch.setBatchId(rs.getLong("batch_id"));
            batch.setNodeId(rs.getString("node_id"));
            batch.setChannelId(rs.getString("channel_id"));
            batch.setStatus(IncomingBatch.Status.valueOf(rs.getString("status")));
            batch.setNetworkMillis(rs.getLong("network_millis"));
            batch.setFilterMillis(rs.getLong("filter_millis"));
            batch.setDatabaseMillis(rs.getLong("database_millis"));
            batch.setFailedRowNumber(rs.getLong("failed_row_number"));
            batch.setFailedLineNumber(rs.getLong("failed_line_number"));
            batch.setByteCount(rs.getLong("byte_count"));
            batch.setStatementCount(rs.getLong("statement_count"));
            batch.setFallbackInsertCount(rs.getLong("fallback_insert_count"));
            batch.setFallbackUpdateCount(rs.getLong("fallback_update_count"));
            batch.setIgnoreCount(rs.getLong("ignore_count"));
            batch.setMissingDeleteCount(rs.getLong("missing_delete_count"));
            batch.setSkipCount(rs.getLong("skip_count"));
            batch.setSqlState(rs.getString("sql_state"));
            batch.setSqlCode(rs.getInt("sql_code"));
            batch.setSqlMessage(rs.getString("sql_message"));
            batch.setLastUpdatedHostName(rs.getString("last_update_hostname"));
            batch.setLastUpdatedTime(rs.getDateTime("last_update_time"));
            batch.setCreateTime(rs.getDateTime("create_time"));
            batch.setErrorFlag(rs.getBoolean("error_flag"));
            return batch;
        }
    }

}

File: symmetric-client-clib/src/service/IncomingBatchService.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "service/IncomingBatchService.h"
#include "common/Log.h"

static SymIncomingBatch * SymIncomingBatchService_mapIncomingBatch(SymRow *row) {
    SymIncomingBatch *batch = SymIncomingBatch_new(NULL);
    batch->batchId = row->getLong(row, "batch_id");
    batch->nodeId = row->getStringNew(row, "node_id");
    batch->channelId = row->getStringNew(row, "channel_id");
    batch->status = row->getStringNew(row, "status");
    batch->networkMillis = row->getInt(row, "network_millis");
    batch->filterMillis = row->getLong(row, "filter_millis");
    batch->databaseMillis = row->getLong(row, "database_millis");
    batch->failedRowNumber = row->getLong(row, "failed_row_number");
    batch->failedLineNumber = row->getLong(row, "failed_line_number");
    batch->byteCount = row->getLong(row, "byte_count");
    batch->statementCount = row->getLong(row, "statement_count");
    batch->fallbackInsertCount = row->getLong(row, "fallback_insert_count");
    batch->fallbackUpdateCount = row->getLong(row, "fallback_update_count");
    batch->ignoreCount = row->getLong(row, "ignore_count");
    batch->missingDeleteCount = row->getLong(row, "missing_delete_count");
    batch->skipCount = row->getLong(row, "skip_count");
    batch->sqlState = row->getStringNew(row, "sql_state");
    batch->sqlCode = row->getInt(row, "sql_code");
    batch->sqlMessage = row->getStringNew(row, "sql_message");
    batch->lastUpdatedHostName = row->getStringNew(row, "last_update_hostname");
    batch->lastUpdatedTime = row->getDate(row, "last_update_time");
    batch->createTime = row->getDate(row, "create_time");
    batch->errorFlag = row->getInt(row, "error_flag");
    return batch;
}

SymIncomingBatch * SymIncomingBatchService_findIncomingBatch(SymIncomingBatchService *this, long batchId, char *nodeId) {
    SymStringBuilder *sql = SymStringBuilder_newWithString(SYM_SQL_SELECT_INCOMING_BATCH_PREFIX);
    sql->append(sql, SYM_SQL_FIND_INCOMING_BATCH);
    SymStringArray *args = SymStringArray_new(NULL);
    args->addLong(args, batchId)->add(args, nodeId);

    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
    int error;
    SymList *batches = sqlTemplate->query(sqlTemplate, sql->str, args, NULL, &error, (void *) SymIncomingBatchService_mapIncomingBatch);
    SymIncomingBatch *incomingBatch = (SymIncomingBatch *) batches->get(batches, 0);
    sql->destroy(sql);
    args->destroy(args);
    batches->destroy(batches);
    return incomingBatch;
}

unsigned short SymIncomingBatchService_isRecordOkBatchesEnabled(SymIncomingBatchService *this) {
    return this->parameterService->is(this->parameterService, SYM_PARAMETER_INCOMING_BATCH_RECORD_OK_ENABLED, 1);
}

unsigned short SymIncomingBatchService_acquireIncomingBatch(SymIncomingBatchService *this, SymIncomingBatch *batch) {
    unsigned short okayToProcess = 1;
    if (batch->batchId > 0) {
        SymIncomingBatch *existingBatch = NULL;

        if (this->isRecordOkBatchesEnabled(this)) {
            existingBatch = this->findIncomingBatch(this, batch->batchId, batch->nodeId);
            if (existingBatch == NULL) {
                this->insertIncomingBatch(this, batch);
            } else {
                batch->retry = 1;
            }
        } else {
            existingBatch = this->findIncomingBatch(this, batch->batchId, batch->nodeId);
            if (existingBatch != NULL) {
                batch->retry = 1;
            }
        }

        if (batch->retry) {
            if (strcmp(existingBatch->status, SYM_INCOMING_BATCH_STATUS_ERROR) == 0
                    || strcmp(existingBatch->status, SYM_INCOMING_BATCH_STATUS_LOADING) == 0
                    || !this->parameterService->is(this->parameterService, SYM_PARAMETER_INCOMING_BATCH_SKIP_DUPLICATE_BATCHES_ENABLED, 1)) {
                okayToProcess = 1;
                existingBatch->status = SYM_INCOMING_BATCH_STATUS_LOADING;
                SymLog_info("Retrying batch %s-%ld", batch->nodeId, batch->batchId);
            } else if (strcmp(existingBatch->status, SYM_INCOMING_BATCH_STATUS_IGNORED) == 0) {
                okayToProcess = 0;
                batch->status = SYM_INCOMING_BATCH_STATUS_OK;
                batch->ignoreCount++;
                existingBatch->status = SYM_INCOMING_BATCH_STATUS_OK;
                existingBatch->ignoreCount++;
                SymLog_info("Ignoring batch %s-%ld", batch->nodeId, batch->batchId);
            } else {
                okayToProcess = 0;
                batch->status = existingBatch->status;
                batch->byteCount = existingBatch->byteCount;
                batch->databaseMillis = existingBatch->databaseMillis;
                batch->networkMillis = existingBatch->networkMillis;
                batch->filterMillis = existingBatch->filterMillis;
                batch->skipCount = existingBatch->skipCount + 1;
                batch->statementCount = existingBatch->statementCount;
                existingBatch->skipCount++;
                SymLog_info("Skipping batch %s-%ld", batch->nodeId, batch->batchId);
            }
            this->updateIncomingBatch(this, existingBatch);
            existingBatch->destroy(existingBatch);
        }
    }
    return okayToProcess;
}

int SymIncomingBatchService_insertIncomingBatch(SymIncomingBatchService *this, SymIncomingBatch *batch) {
    int count = 0;
    if (batch->batchId > 0) {
        batch->lastUpdatedHostName = SymAppUtils_getHostName();
        batch->lastUpdatedTime = SymDate_new();

        SymStringArray *args = SymStringArray_new(NULL);
        args->addLong(args, batch->batchId)->add(args, batch->nodeId)->add(args, batch->channelId);
        args->add(args, batch->status)->addLong(args, batch->networkMillis)->addLong(args, batch->filterMillis);
        args->addLong(args, batch->databaseMillis)->addLong(args, batch->failedRowNumber)->addLong(args, batch->failedLineNumber);
        args->addLong(args, batch->byteCount)->addLong(args, batch->statementCount)->addLong(args, batch->fallbackInsertCount);
        args->addLong(args, batch->fallbackUpdateCount)->addLong(args, batch->ignoreCount)->addLong(args, batch->missingDeleteCount);
        args->addLong(args, batch->skipCount)->add(args, batch->sqlState)->addInt(args, batch->sqlCode);
        args->add(args, batch->sqlMessage)->add(args, batch->lastUpdatedHostName)->add(args, batch->lastUpdatedTime->dateTimeString);

        SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
        int error;
        count = sqlTemplate->update(sqlTemplate, SYM_SQL_INSERT_INCOMING_BATCH, args, NULL, &error);
        args->destroy(args);
    }
    return count;
}

int SymIncomingBatchService_updateIncomingBatch(SymIncomingBatchService *this, SymIncomingBatch *batch) {
    int count = 0;
    if (batch->batchId > 0) {
        if (strcmp(batch->status, SYM_INCOMING_BATCH_STATUS_ERROR) == 0) {
            batch->errorFlag = 1;
        } else if (strcmp(batch->status, SYM_INCOMING_BATCH_STATUS_OK) == 0) {
            batch->errorFlag = 0;
        }
        batch->lastUpdatedHostName = SymAppUtils_getHostName();
        batch->lastUpdatedTime = SymDate_new();

        SymStringArray *args = SymStringArray_new(NULL);
        args->add(args, batch->status)->addInt(args, batch->errorFlag);
        args->addLong(args, batch->networkMillis)->addLong(args, batch->filterMillis);
        args->addLong(args, batch->databaseMillis)->addLong(args, batch->failedRowNumber)->addLong(args, batch->failedLineNumber);
        args->addLong(args, batch->byteCount)->addLong(args, batch->statementCount)->addLong(args, batch->fallbackInsertCount);
        args->addLong(args, batch->fallbackUpdateCount)->addLong(args, batch->ignoreCount)->addLong(args, batch->missingDeleteCount);
        args->addLong(args, batch->skipCount)->add(args, batch->sqlState)->addInt(args, batch->sqlCode);
        args->add(args, batch->sqlMessage)->add(args, batch->lastUpdatedHostName)->add(args, batch->lastUpdatedTime->dateTimeString);
        args->addLong(args, batch->batchId)->add(args, batch->nodeId);

        SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
        int error;
        count = sqlTemplate->update(sqlTemplate, SYM_SQL_UPDATE_INCOMING_BATCH, args, NULL, &error);
        args->destroy(args);
    }
    return count;
}

int SymIncomingBatchService_deleteIncomingBatch(SymIncomingBatchService *this, SymIncomingBatch *batch) {
    SymStringArray *args = SymStringArray_new(NULL);
    args->addLong(args, batch->batchId)->add(args, batch->nodeId);

    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
    int error;
    int count = sqlTemplate->update(sqlTemplate, SYM_SQL_DELETE_INCOMING_BATCH, args, NULL, &error);
    args->destroy(args);
    return count;
}

int SymIncomingBatchService_countIncomingBatchesInError(SymIncomingBatchService *this) {
    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
    int error;
    return sqlTemplate->queryForInt(sqlTemplate, SYM_SQL_COUNT_INCOMING_BATCHES_ERRORS, NULL, NULL, &error);
}

void SymIncomingBatchService_destroy(SymIncomingBatchService *this) {
    free(this);
}

SymIncomingBatchService * SymIncomingBatchService_new(SymIncomingBatchService *this, SymDatabasePlatform *platform, SymParameterService *parameterService) {
    if (this == NULL) {
        this = (SymIncomingBatchService *) calloc(1, sizeof(SymIncomingBatchService));
    }
    this->platform = platform;
    this->parameterService = parameterService;
    this->findIncomingBatch = (void *) &SymIncomingBatchService_findIncomingBatch;
    this->acquireIncomingBatch = (void *) &SymIncomingBatchService_acquireIncomingBatch;
    this->insertIncomingBatch = (void *) &SymIncomingBatchService_insertIncomingBatch;
    this->updateIncomingBatch = (void *) &SymIncomingBatchService_updateIncomingBatch;
    this->deleteIncomingBatch = (void *) &SymIncomingBatchService_deleteIncomingBatch;
    this->isRecordOkBatchesEnabled = (void *) &SymIncomingBatchService_isRecordOkBatchesEnabled;
    this->countIncomingBatchesInError = (void *) &SymIncomingBatchService_countIncomingBatchesInError;
    this->destroy = (void *) &SymIncomingBatchService_destroy;
    return this;
}
